src/Symfony/Bundle/FrameworkBundle/Kernel/MicroKernelTrait.php

```diff
@@ -87,8 +90,10 @@ public function registerContainerConfiguration(LoaderInterface $loader)
                 ],
             ]);
 
+            $kernelClass = str_contains(static::class, "@anonymous\0") ? parent::class : static::class;
```

Comment:
any reason for not using https://www.php.net/manual/en/reflectionclass.isanonymous.php ?

Reply:
because this way saves instanciating an object and is as effective

Reply:
I'd suggest using strpos, that is as effective and saves using the polyfill

Reply:
Mmmh, how strong is that suggestion?

I mean, 
* the polyfill is there anyway
* it should encapsulate the best way to perform this operation on php 7 (or we should fix it)
* on php 8, the native function should be the most efficient way
* I like how it improves readability

Reply:
Relying on php's internal representation rather than publicly exposed api seems a bad solution.. , isn't?

Reply:
> Relying on php's internal representation rather than publicly exposed api seems a bad solution.. , isn't?

I tend to agree. After all, I had to fix quite some code because that internal representation has changed in php 8. Then again, changes like that don't happen that frequently.

Reply:
I don't think strpos is less efficient on php8, but str_contains is on php7 for sure.

About detection logic for anonymous classes, there are tenths of similar checks in the codebase, and many don't fit with reflection at all. This means any change of the internal representation cannot be unnoticed. Using reflection here wouldn't provide anything. It would break consistency of the codebase actually.

Reply:
(updated while merging for strpos)

Reply:
Probably less hacky approach will be:
`assert(get_debug_type(new class {}) === 'class@anonymous');`
Unfortunatetly, it will work only with PHP8

Reply:
@nicolas-grekas In PHP8 `get_debug_type` returns `class@anonymous` for anonymous class objects. I assume that class name prefix is built using this type and may/should remain unchanged in future versions (this question should be addressed to PHP core developers). IMO it is kind of correct to assert anonymous class by its name because they should have some kind static identification.

The only thing I would change:
```php
str_starts_with(static::class, 'class@anonymous');
```

Reply:
> @nicolas-grekas In PHP8 `get_debug_type` returns `class@anonymous` for anonymous class objects.

No, it doesn’t. It’ll give you the parent class instead of just `class` if there is one. Same for `static::class`. The code that you suggested is exactly the kind of code that broke with php 8 all over the codebase.

Reply:
That's an improvement that landed in master, check https://github.com/php/php-src/pull/5153

Reply:
Can someone explain me what is wrong with 

```php
$r = new ReflectionClass(static::class);
$r->isAnonymous();

```
?

Reply:
@goetas `new ReflectionClass($this)` should work as well. :wink: Apart from that: nothing. Your code will achieve the same goal.